### PR TITLE
Initialize regular expressions once

### PR DESCRIPTION
**Pull request by m-kuhn**
## Description

Reduces the cost of preparing static regular expressions repeatedly

Gives a 7% performance boost on simple `like` expressions

### CODE REVIEW

File: src/core/qgsapplication.cpp [hunk 1]
```diff
@@ -1095,7 +1095,8 @@ QString QgsApplication::userStylePath()
 
 QRegExp QgsApplication::shortNameRegExp()
 {
-  return QRegExp( "^[A-Za-z][A-Za-z0-9\\._-]*" );
+  static const QRegExp regexp( QStringLiteral( "^[A-Za-z][A-Za-z0-9\\._-]*" ) );
```
Comment (roya0045) [inline src/core/qgsapplication.cpp:1098]: Does changing this really improve performanc? Or would it be the same as marking the function static const?

Comment (m-kuhn) [reply src/core/qgsapplication.cpp:1098]: The function could be static which is a good idea, but doesn't cache the value automatically.
Making a function static const is not possible afaict.

Comment (roya0045) [reply src/core/qgsapplication.cpp:1098]: Thanks!

Comment (m-kuhn) [reply src/core/qgsapplication.cpp:1098]: Actually it is already static

**Comment by nyalldawson** [comment]
I don't think QRegularExpression is thread safe, is it?

**Comment by m-kuhn** [comment]
good input, on it

**Comment by nyalldawson** [comment]
@m-kuhn one thing you could investigate is if you could leave the static regex, but then detach it before use. Potentially detaching it will be cheaper then creating from scratch....

**Comment by nyalldawson** [comment]
Sorry, I mean copy and detach before use

**Comment by m-kuhn** [comment]
As done here? https://github.com/qgis/QGIS/pull/40737/files#diff-1283eb7f71ac73fd60f9046eeb5340ff665890f26714b5e37e023f8424f5aae5R561-R562

**Comment by nyalldawson** [comment]
No, you have to explicitly detach to make it thread safe. The shallow copy alone isn't sufficient. (See https://stackoverflow.com/questions/25581696/qregularexpression-matching-thread-safety#comment44631377_25583104). But I'm not sure if detaching is the same cost as constructing anew....

**Comment by m-kuhn** [comment]
One more try ... Looking forward to you killing that one too :D

File: src/core/expression/qgsexpressionnodeimpl.cpp [hunk 2]
```diff
@@ -532,39 +532,39 @@ QVariant QgsExpressionNodeBinaryOperator::evalNode( QgsExpression *parent, const
...
           pos = 0;
-          while ( ( pos = rx.indexIn( esc_regexp, pos ) ) != -1 )
+          while ( ( pos = rx3.indexIn( esc_regexp, pos ) ) != -1 )
           {
             esc_regexp.replace( pos + 1, 1, '.' );
             pos += 1;
           }
-          rx.setPattern( QStringLiteral( "\\\\_" ) );
-          esc_regexp.replace( rx, QStringLiteral( "_" ) );
+          const thread_local QRegularExpression rx4( QStringLiteral( "\\\\_" ) );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpressionnodeimpl.cpp:561]: Does this even need to be a regular expression? A standard replace would likely be much faster

File: src/core/qgsapplication.cpp [hunk 3]
```diff
@@ -878,7 +878,7 @@ void QgsApplication::setUITheme( const QString &themeName )
   {
     // apply OS-specific UI scale factor to stylesheet's em values
     int index = 0;
-    QRegularExpression regex( QStringLiteral( "(?<=[\\s:])([0-9\\.]+)(?=em)" ) );
```
Comment (nyalldawson) [inline src/core/qgsapplication.cpp:881]: This one is safe to make static -- no one should be calling setUiTheme from the non gui thread

**Comment by nyalldawson** [comment]
Great work -- the new approach looks safe to me!

**Comment by elpaso** [comment]
Looks good! Two minor improvements that I can recommend: 
1: shouldn't we use QRegularExpression instead of QRegExp in all places (I saw you already replaced some but not all)?
2: Can we use raw strings and get rid of all those unreadable slash escaping?

**Comment by m-kuhn** [comment]
>  1: shouldn't we use QRegularExpression instead of QRegExp in all places (I saw you already replaced some but not all)?

Some had to be reverted, a nice `QRegExp::exactMatch` replacement is only available starting from 5.15, so I propose to keep that for later.

>  2: Can we use raw strings and get rid of all those unreadable slash escaping?

Yes. But I already had to fight hard to get it to the current state, I'd like to keep additional improvements separate

File: src/core/qgscoordinatereferencesystem.cpp [hunk 4]
```diff
@@ -405,8 +405,8 @@ bool QgsCoordinateReferenceSystem::createFromOgcWmsCrs( const QString &crs )
 
   QString wmsCrs = crs;
 
-  QRegExp re_uri( "http://www\\.opengis\\.net/def/crs/([^/]+).+/([^/]+)", Qt::CaseInsensitive );
-  QRegExp re_urn( "urn:ogc:def:crs:([^:]+).+([^:]+)", Qt::CaseInsensitive );
+  thread_local const QRegExp re_uri( QStringLiteral( "http://www\\.opengis\\.net/def/crs/([^/]+).+/([^/]+)" ), Qt::CaseInsensitive );
+  thread_local const QRegExp re_urn( "urn:ogc:def:crs:([^:]+).+([^:]+)", Qt::CaseInsensitive );
```
Comment (m-kuhn) [inline src/core/qgscoordinatereferencesystem.cpp:409]:
```suggestion
  thread_local const QRegExp re_urn( QStringLiteral( "urn:ogc:def:crs:([^:]+).+([^:]+)" ), Qt::CaseInsensitive );
```

File: src/core/qgscoordinatereferencesystem.cpp [hunk 5]
```diff
@@ -316,7 +316,7 @@ bool QgsCoordinateReferenceSystem::createFromString( const QString &definition )
   }
   else
   {
-    QRegularExpression reCrsStr( "^(?:(wkt|proj4|proj)\\:)?(.+)$", QRegularExpression::CaseInsensitiveOption );
+    const thread_local QRegularExpression reCrsStr( "^(?:(wkt|proj4|proj)\\:)?(.+)$", QRegularExpression::CaseInsensitiveOption );
```
Comment (m-kuhn) [inline src/core/qgscoordinatereferencesystem.cpp:319]:
```suggestion
    const thread_local QRegularExpression reCrsStr( QStringLiteral( "^(?:(wkt|proj4|proj)\\:)?(.+)$" ), QRegularExpression::CaseInsensitiveOption );
```

**Pull request merged**